consensus/polybft/state.go

```diff
@@ -51,123 +46,342 @@ const (
+	if res != nil && res.Key()+1 != obj.Key() {
+		// for the case when there is a gap in state syncs
+		return nil, errNotEnoughStateSyncs
 	}
 
-	return newStateSyncEvent(id.Uint64(), sender, target, data, log), nil
+	s.next = obj
+
+	return res, nil
 }
 
-func (s *StateSyncEvent) String() string {
-	return fmt.Sprintf("Id=%d, Sender=%v, Target=%v", s.ID, s.Sender, s.Target)
+// MemDBRecordIterator is a custom iterator for all MemDBRecords
+type MemDBRecordIterator struct {
+	iter       memdb.ResultIterator
+	upperBound uint64
+	next       MemDBRecord
 }
 
-// MessageSignature encapsulates sender identifier and its signature
-type MessageSignature struct {
-	// Signer of the vote
-	From pbft.NodeID
-	// Signature of the message
-	Signature []byte
+// Next returns the next item of the iterator. It is **not** safe to call
+// Next after the last call returned nil.
+func (s *MemDBRecordIterator) Next() (MemDBRecord, error) {
+	res := s.next
+	s.next = nil
+
+	nextItem := s.iter.Next()
+
+	if nextItem == nil {
+		return res, nil
+	}
+
+	obj, isOk := nextItem.(MemDBRecord)
+	if !isOk {
+		return nil, errors.New("item not a MemDBRecord")
+	}
+
+	if obj.Key() > s.upperBound {
+		return res, nil
+	}
+
+	s.next = obj
+
+	return res, nil
 }
 
-// TransportMessage represents the payload which is gossiped across the network
-type TransportMessage struct {
-	// Hash is encoded data
-	Hash []byte
-	// Message signature
-	Signature []byte
-	// Node identifier
-	NodeID pbft.NodeID
-	// Number of epoch
-	EpochNumber uint64
+// insertToMemDB inserts a new record in provided table
+func insertToMemDB[V MemDBRecord](memdb *memdb.MemDB, table string, record V) error {
+	txn := memdb.Txn(true)
+	if err := txn.Insert(table, record); err != nil {
+		txn.Abort()
+
+		return err
+	}
+
+	txn.Commit()
+
+	return nil
 }
 
-var (
-	// bucket to store rootchain bridge events
-	syncStateEventsBucket = []byte("events")
-	//bucket to store commitments
-	commitmentsBucket = []byte("commitments")
-	// bucket to store bundles
-	bundlesBucket = []byte("bundles")
-	// bucket to store epochs and all its nested buckets (message votes and message pool events)
-	epochsBucket = []byte("epochs")
-	// bucket to store message votes (signatures)
-	messageVotesBucket = []byte("votes")
-	// bucket to store validator snapshots
-	validatorSnapshotsBucket = []byte("validatorSnapshots")
-	// array of all parent buckets
-	parentBuckets = [][]byte{syncStateEventsBucket, commitmentsBucket, bundlesBucket,
-		epochsBucket, validatorSnapshotsBucket}
-	// ErrNotEnoughStateSyncs error message
-	ErrNotEnoughStateSyncs = errors.New("there is either a gap or not enough sync events")
-	// ErrCommitmentNotBuilt error message
-	ErrCommitmentNotBuilt = errors.New("there is no built commitment to register")
-)
+// getFilteredFromMemDB returns a filtered collection of desired memdb records
+func getFilteredFromMemDB[V MemDBRecord](memdb *memdb.MemDB, table string,
+	lowerBound, upperBound interface{}) ([]V, error) {
```

Comment:
why do we want to use lowerBound and upperBound as empty interfaces?

Reply:
`lowerBound, upperBound interface{}` - why not to have them as generics also? like numeric type?

Reply:
Yep, you are right. I'll change that to be generics as well.

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/pull/785/commits/7b91a8734e719e42fac631fa4abd0bd629ae04ce